{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview of GTEx Resources: working with gene expression, eQTL, and metadata files\n",
    "\n",
    "\n",
    "This notebook illustrates how to parse the GTEx data tables and resources available from the [GTEx Portal](https://gtexportal.org), including:\n",
    "* Sample and subject attributes\n",
    "* Expression tables\n",
    "* eQTL results\n",
    "\n",
    "All examples use data from the V7 release.\n",
    "\n",
    "#### Pre-requisites\n",
    "\n",
    "**_For the online version of this notebook, all datasets used in the examples are available to the notebook, and no additional files are required._**\n",
    "\n",
    "To run the notebook elsewhere, the following files are needed (all are available from the [downloads](https://gtexportal.org/home/datasets) section of the portal):\n",
    "* GTEx_v7_Annotations_SampleAttributesDS.txt\n",
    "* GTEx_v7_Annotations_SubjectPhenotypesDS.txt\n",
    "* GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.gct.gz\n",
    "* GTEx_Analysis_v7_eQTL.tar.gz\n",
    "\n",
    "Uncompress GTEx_Analysis_v7_eQTL.tar.gz, and make sure that the contents of the resulting GTEx_Analysis_v7_eQTL directory are in the same location as the other files. In the code below, it is assumed that all data is stored in a `data` directory, at the same level as this notebook.\n",
    "\n",
    "#### Python environment\n",
    "\n",
    "The `pandas` library is used to read and parse the data. Run the following import statements to set up the environment:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selecting samples and subsets of the data\n",
    "\n",
    "To begin, we will explore the sample and subject annotations. Load the two tables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_df = pd.read_csv('data/GTEx_v7_Annotations_SampleAttributesDS.txt', sep='\\t', index_col=0)\n",
    "subject_df = pd.read_csv('data/GTEx_v7_Annotations_SubjectPhenotypesDS.txt', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Explanations for each column header can be found in the dictionary spreadsheets available on the portal (provided here for convenience):<br>\n",
    "[GTEx_Analysis_v7_Annotations_SubjectPhenotypesDD.xlsx](data/GTEx_Analysis_v7_Annotations_SubjectPhenotypesDD.xlsx)<br>\n",
    "[GTEx_Analysis_v7_Annotations_SampleAttributesDD.xlsx](data/GTEx_Analysis_v7_Annotations_SampleAttributesDD.xlsx)\n",
    "\n",
    "#### Subject attributes\n",
    "\n",
    "To view the first few rows of the subject attributes, run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE</th>\n",
       "      <th>DTHHRDY</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SUBJID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <td>2</td>\n",
       "      <td>60-69</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111CU</th>\n",
       "      <td>1</td>\n",
       "      <td>50-59</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <td>1</td>\n",
       "      <td>60-69</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111VG</th>\n",
       "      <td>1</td>\n",
       "      <td>60-69</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111YS</th>\n",
       "      <td>1</td>\n",
       "      <td>60-69</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            SEX    AGE  DTHHRDY\n",
       "SUBJID                         \n",
       "GTEX-1117F    2  60-69      4.0\n",
       "GTEX-111CU    1  50-59      0.0\n",
       "GTEX-111FC    1  60-69      1.0\n",
       "GTEX-111VG    1  60-69      3.0\n",
       "GTEX-111YS    1  60-69      0.0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subject_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `DTHHRDY` column encodes the death classification based on the 4-point _Hardy Scale_. The coding corresponds to the following categories:<br>\n",
    "0: Ventilator Case; 1: Violent and fast death; 2: Fast death of natural causes; 3: Intermediate death; 4: Slow death\n",
    "\n",
    "For additional details on the subject attributes, see [GTEx_Analysis_v7_Annotations_SubjectPhenotypesDD.xlsx](data/GTEx_Analysis_v7_Annotations_SubjectPhenotypesDD.xlsx).\n",
    "\n",
    "_Note that the attribute tables from the GTEx Portal contain open-access data only; protected information (e.g., exact donor age, phenotypic information, etc.) is available from dbGaP._ Consequently, only age ranges are provided here.\n",
    "\n",
    "#### Sample attributes\n",
    "\n",
    "The sample attributes file contains information on all samples in the GTEx resource, including RNA-seq samples and whole genome sequence (WGS) samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMATSSCR</th>\n",
       "      <th>SMCENTER</th>\n",
       "      <th>SMPTHNTS</th>\n",
       "      <th>SMRIN</th>\n",
       "      <th>SMTS</th>\n",
       "      <th>SMTSD</th>\n",
       "      <th>SMUBRID</th>\n",
       "      <th>SMTSISCH</th>\n",
       "      <th>SMTSPAX</th>\n",
       "      <th>SMNABTCH</th>\n",
       "      <th>...</th>\n",
       "      <th>SME1ANTI</th>\n",
       "      <th>SMSPLTRD</th>\n",
       "      <th>SMBSMMRT</th>\n",
       "      <th>SME1SNSE</th>\n",
       "      <th>SME1PCTS</th>\n",
       "      <th>SMRRNART</th>\n",
       "      <th>SME1MPRT</th>\n",
       "      <th>SMNUM5CD</th>\n",
       "      <th>SMDPMPRT</th>\n",
       "      <th>SME2PCTS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-58Q7G</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "      <td>1188.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BP-38516</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-5DWSB</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "      <td>1188.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BP-38516</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0003-SM-6WBT7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>B1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Blood</td>\n",
       "      <td>Whole Blood</td>\n",
       "      <td>0013756</td>\n",
       "      <td>1188.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BP-38543</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, ~15% vessel stroma, rep delineated</td>\n",
       "      <td>6.8</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>0002190</td>\n",
       "      <td>1214.0</td>\n",
       "      <td>1125.0</td>\n",
       "      <td>BP-43693</td>\n",
       "      <td>...</td>\n",
       "      <td>14579275.0</td>\n",
       "      <td>12025354.0</td>\n",
       "      <td>0.003164</td>\n",
       "      <td>14634407.0</td>\n",
       "      <td>50.094357</td>\n",
       "      <td>0.003102</td>\n",
       "      <td>0.992826</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.12628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>0.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, !5% fibrous connective tissue, delin...</td>\n",
       "      <td>7.1</td>\n",
       "      <td>Muscle</td>\n",
       "      <td>Muscle - Skeletal</td>\n",
       "      <td>0011907</td>\n",
       "      <td>1220.0</td>\n",
       "      <td>1119.0</td>\n",
       "      <td>BP-43495</td>\n",
       "      <td>...</td>\n",
       "      <td>13134349.0</td>\n",
       "      <td>11578874.0</td>\n",
       "      <td>0.003991</td>\n",
       "      <td>13307871.0</td>\n",
       "      <td>50.328114</td>\n",
       "      <td>0.006991</td>\n",
       "      <td>0.994212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.90517</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SMATSSCR SMCENTER  \\\n",
       "SAMPID                                        \n",
       "GTEX-1117F-0003-SM-58Q7G       NaN       B1   \n",
       "GTEX-1117F-0003-SM-5DWSB       NaN       B1   \n",
       "GTEX-1117F-0003-SM-6WBT7       NaN       B1   \n",
       "GTEX-1117F-0226-SM-5GZZ7       0.0       B1   \n",
       "GTEX-1117F-0426-SM-5EGHI       0.0       B1   \n",
       "\n",
       "                                                                   SMPTHNTS  \\\n",
       "SAMPID                                                                        \n",
       "GTEX-1117F-0003-SM-58Q7G                                                NaN   \n",
       "GTEX-1117F-0003-SM-5DWSB                                                NaN   \n",
       "GTEX-1117F-0003-SM-6WBT7                                                NaN   \n",
       "GTEX-1117F-0226-SM-5GZZ7       2 pieces, ~15% vessel stroma, rep delineated   \n",
       "GTEX-1117F-0426-SM-5EGHI  2 pieces, !5% fibrous connective tissue, delin...   \n",
       "\n",
       "                          SMRIN            SMTS                   SMTSD  \\\n",
       "SAMPID                                                                    \n",
       "GTEX-1117F-0003-SM-58Q7G    NaN           Blood             Whole Blood   \n",
       "GTEX-1117F-0003-SM-5DWSB    NaN           Blood             Whole Blood   \n",
       "GTEX-1117F-0003-SM-6WBT7    NaN           Blood             Whole Blood   \n",
       "GTEX-1117F-0226-SM-5GZZ7    6.8  Adipose Tissue  Adipose - Subcutaneous   \n",
       "GTEX-1117F-0426-SM-5EGHI    7.1          Muscle       Muscle - Skeletal   \n",
       "\n",
       "                          SMUBRID  SMTSISCH  SMTSPAX  SMNABTCH    ...     \\\n",
       "SAMPID                                                            ...      \n",
       "GTEX-1117F-0003-SM-58Q7G  0013756    1188.0      NaN  BP-38516    ...      \n",
       "GTEX-1117F-0003-SM-5DWSB  0013756    1188.0      NaN  BP-38516    ...      \n",
       "GTEX-1117F-0003-SM-6WBT7  0013756    1188.0      NaN  BP-38543    ...      \n",
       "GTEX-1117F-0226-SM-5GZZ7  0002190    1214.0   1125.0  BP-43693    ...      \n",
       "GTEX-1117F-0426-SM-5EGHI  0011907    1220.0   1119.0  BP-43495    ...      \n",
       "\n",
       "                            SME1ANTI    SMSPLTRD  SMBSMMRT    SME1SNSE  \\\n",
       "SAMPID                                                                   \n",
       "GTEX-1117F-0003-SM-58Q7G         NaN         NaN       NaN         NaN   \n",
       "GTEX-1117F-0003-SM-5DWSB         NaN         NaN       NaN         NaN   \n",
       "GTEX-1117F-0003-SM-6WBT7         NaN         NaN       NaN         NaN   \n",
       "GTEX-1117F-0226-SM-5GZZ7  14579275.0  12025354.0  0.003164  14634407.0   \n",
       "GTEX-1117F-0426-SM-5EGHI  13134349.0  11578874.0  0.003991  13307871.0   \n",
       "\n",
       "                           SME1PCTS  SMRRNART  SME1MPRT  SMNUM5CD  SMDPMPRT  \\\n",
       "SAMPID                                                                        \n",
       "GTEX-1117F-0003-SM-58Q7G        NaN       NaN       NaN       NaN       NaN   \n",
       "GTEX-1117F-0003-SM-5DWSB        NaN       NaN       NaN       NaN       NaN   \n",
       "GTEX-1117F-0003-SM-6WBT7        NaN       NaN       NaN       NaN       NaN   \n",
       "GTEX-1117F-0226-SM-5GZZ7  50.094357  0.003102  0.992826       NaN       0.0   \n",
       "GTEX-1117F-0426-SM-5EGHI  50.328114  0.006991  0.994212       NaN       0.0   \n",
       "\n",
       "                          SME2PCTS  \n",
       "SAMPID                              \n",
       "GTEX-1117F-0003-SM-58Q7G       NaN  \n",
       "GTEX-1117F-0003-SM-5DWSB       NaN  \n",
       "GTEX-1117F-0003-SM-6WBT7       NaN  \n",
       "GTEX-1117F-0226-SM-5GZZ7  50.12628  \n",
       "GTEX-1117F-0426-SM-5EGHI  49.90517  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each GTEx release defines an analysis freeze containing samples that pass all quality controls, for each sample type. The analysis freeze column is called `SMAFRZE`. For a summary of analysis freeze samples of each data type in the V7 release, run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RNASEQ     11688\n",
       "EXCLUDE     2232\n",
       "WGS          635\n",
       "WES          593\n",
       "OMNI         450\n",
       "Name: SMAFRZE, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df['SMAFRZE'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we'll focus on the RNA-seq samples. To facilitate this, we define a new data frame containing the RNA-seq analysis freeze samples only:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnaseq_sample_df = sample_df[sample_df['SMAFRZE']=='RNASEQ']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To view the first few rows of attributes for these samples, run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMATSSCR</th>\n",
       "      <th>SMCENTER</th>\n",
       "      <th>SMPTHNTS</th>\n",
       "      <th>SMRIN</th>\n",
       "      <th>SMTS</th>\n",
       "      <th>SMTSD</th>\n",
       "      <th>SMUBRID</th>\n",
       "      <th>SMTSISCH</th>\n",
       "      <th>SMTSPAX</th>\n",
       "      <th>SMNABTCH</th>\n",
       "      <th>...</th>\n",
       "      <th>SME1ANTI</th>\n",
       "      <th>SMSPLTRD</th>\n",
       "      <th>SMBSMMRT</th>\n",
       "      <th>SME1SNSE</th>\n",
       "      <th>SME1PCTS</th>\n",
       "      <th>SMRRNART</th>\n",
       "      <th>SME1MPRT</th>\n",
       "      <th>SMNUM5CD</th>\n",
       "      <th>SMDPMPRT</th>\n",
       "      <th>SME2PCTS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>0.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, ~15% vessel stroma, rep delineated</td>\n",
       "      <td>6.8</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>0002190</td>\n",
       "      <td>1214.0</td>\n",
       "      <td>1125.0</td>\n",
       "      <td>BP-43693</td>\n",
       "      <td>...</td>\n",
       "      <td>14579275.0</td>\n",
       "      <td>12025354.0</td>\n",
       "      <td>0.003164</td>\n",
       "      <td>14634407.0</td>\n",
       "      <td>50.094357</td>\n",
       "      <td>0.003102</td>\n",
       "      <td>0.992826</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.126280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>0.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, !5% fibrous connective tissue, delin...</td>\n",
       "      <td>7.1</td>\n",
       "      <td>Muscle</td>\n",
       "      <td>Muscle - Skeletal</td>\n",
       "      <td>0011907</td>\n",
       "      <td>1220.0</td>\n",
       "      <td>1119.0</td>\n",
       "      <td>BP-43495</td>\n",
       "      <td>...</td>\n",
       "      <td>13134349.0</td>\n",
       "      <td>11578874.0</td>\n",
       "      <td>0.003991</td>\n",
       "      <td>13307871.0</td>\n",
       "      <td>50.328114</td>\n",
       "      <td>0.006991</td>\n",
       "      <td>0.994212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.905170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <td>0.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, clean, Monckebeg medial sclerosis, r...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>Blood Vessel</td>\n",
       "      <td>Artery - Tibial</td>\n",
       "      <td>0007610</td>\n",
       "      <td>1221.0</td>\n",
       "      <td>1120.0</td>\n",
       "      <td>BP-43495</td>\n",
       "      <td>...</td>\n",
       "      <td>13169835.0</td>\n",
       "      <td>11015113.0</td>\n",
       "      <td>0.004285</td>\n",
       "      <td>13160068.0</td>\n",
       "      <td>49.981450</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.992711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.227848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <td>1.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, up to 4mm aderent fat/nerve/vessel, ...</td>\n",
       "      <td>6.9</td>\n",
       "      <td>Blood Vessel</td>\n",
       "      <td>Artery - Coronary</td>\n",
       "      <td>0001621</td>\n",
       "      <td>1243.0</td>\n",
       "      <td>1098.0</td>\n",
       "      <td>BP-43956</td>\n",
       "      <td>...</td>\n",
       "      <td>15148343.0</td>\n",
       "      <td>11624467.0</td>\n",
       "      <td>0.003379</td>\n",
       "      <td>15282444.0</td>\n",
       "      <td>50.220333</td>\n",
       "      <td>0.005335</td>\n",
       "      <td>0.991175</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.025043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <td>1.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, no abnormalities</td>\n",
       "      <td>6.3</td>\n",
       "      <td>Heart</td>\n",
       "      <td>Heart - Atrial Appendage</td>\n",
       "      <td>0006631</td>\n",
       "      <td>1244.0</td>\n",
       "      <td>1097.0</td>\n",
       "      <td>BP-44261</td>\n",
       "      <td>...</td>\n",
       "      <td>13583226.0</td>\n",
       "      <td>9262806.0</td>\n",
       "      <td>0.003451</td>\n",
       "      <td>13745609.0</td>\n",
       "      <td>50.297090</td>\n",
       "      <td>0.030579</td>\n",
       "      <td>0.994478</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.929870</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SMATSSCR SMCENTER  \\\n",
       "SAMPID                                        \n",
       "GTEX-1117F-0226-SM-5GZZ7       0.0       B1   \n",
       "GTEX-1117F-0426-SM-5EGHI       0.0       B1   \n",
       "GTEX-1117F-0526-SM-5EGHJ       0.0       B1   \n",
       "GTEX-1117F-0626-SM-5N9CS       1.0       B1   \n",
       "GTEX-1117F-0726-SM-5GIEN       1.0       B1   \n",
       "\n",
       "                                                                   SMPTHNTS  \\\n",
       "SAMPID                                                                        \n",
       "GTEX-1117F-0226-SM-5GZZ7       2 pieces, ~15% vessel stroma, rep delineated   \n",
       "GTEX-1117F-0426-SM-5EGHI  2 pieces, !5% fibrous connective tissue, delin...   \n",
       "GTEX-1117F-0526-SM-5EGHJ  2 pieces, clean, Monckebeg medial sclerosis, r...   \n",
       "GTEX-1117F-0626-SM-5N9CS  2 pieces, up to 4mm aderent fat/nerve/vessel, ...   \n",
       "GTEX-1117F-0726-SM-5GIEN                         2 pieces, no abnormalities   \n",
       "\n",
       "                          SMRIN            SMTS                     SMTSD  \\\n",
       "SAMPID                                                                      \n",
       "GTEX-1117F-0226-SM-5GZZ7    6.8  Adipose Tissue    Adipose - Subcutaneous   \n",
       "GTEX-1117F-0426-SM-5EGHI    7.1          Muscle         Muscle - Skeletal   \n",
       "GTEX-1117F-0526-SM-5EGHJ    8.0    Blood Vessel           Artery - Tibial   \n",
       "GTEX-1117F-0626-SM-5N9CS    6.9    Blood Vessel         Artery - Coronary   \n",
       "GTEX-1117F-0726-SM-5GIEN    6.3           Heart  Heart - Atrial Appendage   \n",
       "\n",
       "                          SMUBRID  SMTSISCH  SMTSPAX  SMNABTCH    ...      \\\n",
       "SAMPID                                                            ...       \n",
       "GTEX-1117F-0226-SM-5GZZ7  0002190    1214.0   1125.0  BP-43693    ...       \n",
       "GTEX-1117F-0426-SM-5EGHI  0011907    1220.0   1119.0  BP-43495    ...       \n",
       "GTEX-1117F-0526-SM-5EGHJ  0007610    1221.0   1120.0  BP-43495    ...       \n",
       "GTEX-1117F-0626-SM-5N9CS  0001621    1243.0   1098.0  BP-43956    ...       \n",
       "GTEX-1117F-0726-SM-5GIEN  0006631    1244.0   1097.0  BP-44261    ...       \n",
       "\n",
       "                            SME1ANTI    SMSPLTRD  SMBSMMRT    SME1SNSE  \\\n",
       "SAMPID                                                                   \n",
       "GTEX-1117F-0226-SM-5GZZ7  14579275.0  12025354.0  0.003164  14634407.0   \n",
       "GTEX-1117F-0426-SM-5EGHI  13134349.0  11578874.0  0.003991  13307871.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ  13169835.0  11015113.0  0.004285  13160068.0   \n",
       "GTEX-1117F-0626-SM-5N9CS  15148343.0  11624467.0  0.003379  15282444.0   \n",
       "GTEX-1117F-0726-SM-5GIEN  13583226.0   9262806.0  0.003451  13745609.0   \n",
       "\n",
       "                           SME1PCTS  SMRRNART  SME1MPRT  SMNUM5CD  SMDPMPRT  \\\n",
       "SAMPID                                                                        \n",
       "GTEX-1117F-0226-SM-5GZZ7  50.094357  0.003102  0.992826       NaN       0.0   \n",
       "GTEX-1117F-0426-SM-5EGHI  50.328114  0.006991  0.994212       NaN       0.0   \n",
       "GTEX-1117F-0526-SM-5EGHJ  49.981450  0.002867  0.992711       NaN       0.0   \n",
       "GTEX-1117F-0626-SM-5N9CS  50.220333  0.005335  0.991175       NaN       0.0   \n",
       "GTEX-1117F-0726-SM-5GIEN  50.297090  0.030579  0.994478       NaN       0.0   \n",
       "\n",
       "                           SME2PCTS  \n",
       "SAMPID                               \n",
       "GTEX-1117F-0226-SM-5GZZ7  50.126280  \n",
       "GTEX-1117F-0426-SM-5EGHI  49.905170  \n",
       "GTEX-1117F-0526-SM-5EGHJ  50.227848  \n",
       "GTEX-1117F-0626-SM-5N9CS  50.025043  \n",
       "GTEX-1117F-0726-SM-5GIEN  49.929870  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnaseq_sample_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One of the key columns in the sample attribute table is `SMTSD`, which stands for 'Tissue Site Detail'. This enables the selection of samples from a specific tissue. For example, to obtain a summary of sample sizes in each tissue, run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Muscle - Skeletal                            564\n",
       "Skin - Sun Exposed (Lower leg)               473\n",
       "Thyroid                                      446\n",
       "Adipose - Subcutaneous                       442\n",
       "Artery - Tibial                              441\n",
       "Lung                                         427\n",
       "Nerve - Tibial                               414\n",
       "Esophagus - Mucosa                           407\n",
       "Whole Blood                                  407\n",
       "Skin - Not Sun Exposed (Suprapubic)          387\n",
       "Esophagus - Muscularis                       370\n",
       "Adipose - Visceral (Omentum)                 355\n",
       "Cells - Transformed fibroblasts              343\n",
       "Heart - Left Ventricle                       303\n",
       "Artery - Aorta                               299\n",
       "Heart - Atrial Appendage                     297\n",
       "Breast - Mammary Tissue                      290\n",
       "Colon - Transverse                           274\n",
       "Stomach                                      262\n",
       "Testis                                       259\n",
       "Pancreas                                     248\n",
       "Esophagus - Gastroesophageal Junction        244\n",
       "Colon - Sigmoid                              233\n",
       "Adrenal Gland                                190\n",
       "Pituitary                                    183\n",
       "Liver                                        175\n",
       "Artery - Coronary                            173\n",
       "Brain - Cerebellum                           173\n",
       "Spleen                                       162\n",
       "Brain - Caudate (basal ganglia)              160\n",
       "Brain - Cortex                               158\n",
       "Prostate                                     152\n",
       "Brain - Nucleus accumbens (basal ganglia)    147\n",
       "Small Intestine - Terminal Ileum             137\n",
       "Brain - Cerebellar Hemisphere                136\n",
       "Ovary                                        133\n",
       "Cells - EBV-transformed lymphocytes          130\n",
       "Brain - Frontal Cortex (BA9)                 129\n",
       "Brain - Putamen (basal ganglia)              124\n",
       "Brain - Hippocampus                          123\n",
       "Brain - Hypothalamus                         121\n",
       "Brain - Anterior cingulate cortex (BA24)     121\n",
       "Vagina                                       115\n",
       "Uterus                                       111\n",
       "Brain - Amygdala                             100\n",
       "Minor Salivary Gland                          97\n",
       "Brain - Spinal cord (cervical c-1)            91\n",
       "Brain - Substantia nigra                      88\n",
       "Kidney - Cortex                               45\n",
       "Bladder                                       11\n",
       "Fallopian Tube                                 7\n",
       "Cervix - Ectocervix                            6\n",
       "Cervix - Endocervix                            5\n",
       "Name: SMTSD, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnaseq_sample_df['SMTSD'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The RNA-seq samples corresponding to a given tissue can be selected using the `SMTSD` column:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SMATSSCR</th>\n",
       "      <th>SMCENTER</th>\n",
       "      <th>SMPTHNTS</th>\n",
       "      <th>SMRIN</th>\n",
       "      <th>SMTS</th>\n",
       "      <th>SMTSD</th>\n",
       "      <th>SMUBRID</th>\n",
       "      <th>SMTSISCH</th>\n",
       "      <th>SMTSPAX</th>\n",
       "      <th>SMNABTCH</th>\n",
       "      <th>...</th>\n",
       "      <th>SME1ANTI</th>\n",
       "      <th>SMSPLTRD</th>\n",
       "      <th>SMBSMMRT</th>\n",
       "      <th>SME1SNSE</th>\n",
       "      <th>SME1PCTS</th>\n",
       "      <th>SMRRNART</th>\n",
       "      <th>SME1MPRT</th>\n",
       "      <th>SMNUM5CD</th>\n",
       "      <th>SMDPMPRT</th>\n",
       "      <th>SME2PCTS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-111CU-0526-SM-5EGHK</th>\n",
       "      <td>1.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, include 20 and 40% fat (attached and...</td>\n",
       "      <td>7.2</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>0001150</td>\n",
       "      <td>72.0</td>\n",
       "      <td>637.0</td>\n",
       "      <td>BP-43495</td>\n",
       "      <td>...</td>\n",
       "      <td>13106749.0</td>\n",
       "      <td>15136415.0</td>\n",
       "      <td>0.008059</td>\n",
       "      <td>13257128.0</td>\n",
       "      <td>50.285200</td>\n",
       "      <td>0.001842</td>\n",
       "      <td>0.992940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.936325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111YS-1226-SM-5EGGJ</th>\n",
       "      <td>1.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 somewhat fragmented pieces</td>\n",
       "      <td>6.8</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>0001150</td>\n",
       "      <td>154.0</td>\n",
       "      <td>735.0</td>\n",
       "      <td>BP-44460</td>\n",
       "      <td>...</td>\n",
       "      <td>15480965.0</td>\n",
       "      <td>20021609.0</td>\n",
       "      <td>0.005362</td>\n",
       "      <td>15599862.0</td>\n",
       "      <td>50.191270</td>\n",
       "      <td>0.003316</td>\n",
       "      <td>0.993188</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.074604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1122O-0726-SM-5GIEV</th>\n",
       "      <td>1.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, minimal attached and internal fat</td>\n",
       "      <td>6.0</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>0001150</td>\n",
       "      <td>55.0</td>\n",
       "      <td>948.0</td>\n",
       "      <td>BP-44261</td>\n",
       "      <td>...</td>\n",
       "      <td>21587613.0</td>\n",
       "      <td>25147179.0</td>\n",
       "      <td>0.006104</td>\n",
       "      <td>21545407.0</td>\n",
       "      <td>49.951077</td>\n",
       "      <td>0.002716</td>\n",
       "      <td>0.993875</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.221860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1128S-0826-SM-5GZZI</th>\n",
       "      <td>1.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces, up to 50% interstital fat, rep delin...</td>\n",
       "      <td>6.2</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>0001150</td>\n",
       "      <td>842.0</td>\n",
       "      <td>1003.0</td>\n",
       "      <td>BP-43753</td>\n",
       "      <td>...</td>\n",
       "      <td>18722483.0</td>\n",
       "      <td>20292290.0</td>\n",
       "      <td>0.005499</td>\n",
       "      <td>18640983.0</td>\n",
       "      <td>49.890940</td>\n",
       "      <td>0.004870</td>\n",
       "      <td>0.993642</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.145958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-117YX-0226-SM-5EGH6</th>\n",
       "      <td>1.0</td>\n",
       "      <td>B1</td>\n",
       "      <td>2 pieces; well dissected, no fat</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>0001150</td>\n",
       "      <td>80.0</td>\n",
       "      <td>726.0</td>\n",
       "      <td>BP-44460</td>\n",
       "      <td>...</td>\n",
       "      <td>15319276.0</td>\n",
       "      <td>18135569.0</td>\n",
       "      <td>0.007021</td>\n",
       "      <td>15453346.0</td>\n",
       "      <td>50.217842</td>\n",
       "      <td>0.004132</td>\n",
       "      <td>0.992800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.853798</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SMATSSCR SMCENTER  \\\n",
       "SAMPID                                        \n",
       "GTEX-111CU-0526-SM-5EGHK       1.0       B1   \n",
       "GTEX-111YS-1226-SM-5EGGJ       1.0       B1   \n",
       "GTEX-1122O-0726-SM-5GIEV       1.0       B1   \n",
       "GTEX-1128S-0826-SM-5GZZI       1.0       B1   \n",
       "GTEX-117YX-0226-SM-5EGH6       1.0       B1   \n",
       "\n",
       "                                                                   SMPTHNTS  \\\n",
       "SAMPID                                                                        \n",
       "GTEX-111CU-0526-SM-5EGHK  2 pieces, include 20 and 40% fat (attached and...   \n",
       "GTEX-111YS-1226-SM-5EGGJ                       2 somewhat fragmented pieces   \n",
       "GTEX-1122O-0726-SM-5GIEV        2 pieces, minimal attached and internal fat   \n",
       "GTEX-1128S-0826-SM-5GZZI  2 pieces, up to 50% interstital fat, rep delin...   \n",
       "GTEX-117YX-0226-SM-5EGH6                   2 pieces; well dissected, no fat   \n",
       "\n",
       "                          SMRIN      SMTS     SMTSD  SMUBRID  SMTSISCH  \\\n",
       "SAMPID                                                                   \n",
       "GTEX-111CU-0526-SM-5EGHK    7.2  Pancreas  Pancreas  0001150      72.0   \n",
       "GTEX-111YS-1226-SM-5EGGJ    6.8  Pancreas  Pancreas  0001150     154.0   \n",
       "GTEX-1122O-0726-SM-5GIEV    6.0  Pancreas  Pancreas  0001150      55.0   \n",
       "GTEX-1128S-0826-SM-5GZZI    6.2  Pancreas  Pancreas  0001150     842.0   \n",
       "GTEX-117YX-0226-SM-5EGH6    7.0  Pancreas  Pancreas  0001150      80.0   \n",
       "\n",
       "                          SMTSPAX  SMNABTCH    ...        SME1ANTI  \\\n",
       "SAMPID                                         ...                   \n",
       "GTEX-111CU-0526-SM-5EGHK    637.0  BP-43495    ...      13106749.0   \n",
       "GTEX-111YS-1226-SM-5EGGJ    735.0  BP-44460    ...      15480965.0   \n",
       "GTEX-1122O-0726-SM-5GIEV    948.0  BP-44261    ...      21587613.0   \n",
       "GTEX-1128S-0826-SM-5GZZI   1003.0  BP-43753    ...      18722483.0   \n",
       "GTEX-117YX-0226-SM-5EGH6    726.0  BP-44460    ...      15319276.0   \n",
       "\n",
       "                            SMSPLTRD  SMBSMMRT    SME1SNSE   SME1PCTS  \\\n",
       "SAMPID                                                                  \n",
       "GTEX-111CU-0526-SM-5EGHK  15136415.0  0.008059  13257128.0  50.285200   \n",
       "GTEX-111YS-1226-SM-5EGGJ  20021609.0  0.005362  15599862.0  50.191270   \n",
       "GTEX-1122O-0726-SM-5GIEV  25147179.0  0.006104  21545407.0  49.951077   \n",
       "GTEX-1128S-0826-SM-5GZZI  20292290.0  0.005499  18640983.0  49.890940   \n",
       "GTEX-117YX-0226-SM-5EGH6  18135569.0  0.007021  15453346.0  50.217842   \n",
       "\n",
       "                          SMRRNART  SME1MPRT  SMNUM5CD  SMDPMPRT   SME2PCTS  \n",
       "SAMPID                                                                       \n",
       "GTEX-111CU-0526-SM-5EGHK  0.001842  0.992940       NaN       0.0  49.936325  \n",
       "GTEX-111YS-1226-SM-5EGGJ  0.003316  0.993188       NaN       0.0  50.074604  \n",
       "GTEX-1122O-0726-SM-5GIEV  0.002716  0.993875       NaN       0.0  50.221860  \n",
       "GTEX-1128S-0826-SM-5GZZI  0.004870  0.993642       NaN       0.0  50.145958  \n",
       "GTEX-117YX-0226-SM-5EGH6  0.004132  0.992800       NaN       0.0  49.853798  \n",
       "\n",
       "[5 rows x 62 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pancreas_sample_df = rnaseq_sample_df[rnaseq_sample_df['SMTSD']=='Pancreas']\n",
    "pancreas_sample_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All GTEx samples are indexed using the same identifier format. Importantly, the first two components of a *sample ID* correspond to the *donor ID* for that sample. For example, to obtain the donor IDs corresponding to the samples selected above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GTEX-111CU-0526-SM-5EGHK',\n",
       " 'GTEX-111YS-1226-SM-5EGGJ',\n",
       " 'GTEX-1122O-0726-SM-5GIEV']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pancreas_sample_df.index[:3].tolist()  # show the first 3 sample IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "run the following:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GTEX-111CU', 'GTEX-111YS', 'GTEX-1122O']\n"
     ]
    }
   ],
   "source": [
    "donor_ids = pancreas_sample_df.index.map(lambda x: '-'.join(x.split('-')[:2]))\n",
    "print(donor_ids[:3].tolist())  # show the first 3 donor IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Samples can then be split by donor attribute as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pancreas_male_df = pancreas_sample_df.loc[subject_df.loc[donor_ids, 'SEX'].values==1]  # males are encoded as 1\n",
    "pancreas_female_df = pancreas_sample_df.loc[subject_df.loc[donor_ids, 'SEX'].values==2]  # females are encoded as 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading expression data\n",
    "Next, after having selected a specific set of samples, we show how to load the expression data corresponding to these samples. Gene-level expression is provided as raw read counts and in TPM (transcripts per million); here we will use the TPM data. Both tables are in GCT format, which contains two rows of header data. The first two columns contain the gene ID and gene name; since gene names are not unique, **all GTEx tables are indexed by GENCODE/ENSEMBL gene ID**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes in table: 56202\n"
     ]
    }
   ],
   "source": [
    "# to select a specific set of samples, pass the IDs to 'read_csv':\n",
    "pancreas_sample_ids = ['Name']+pancreas_sample_df.index.tolist()  # add 'Name' to load ENSEMBLE ID of the gene\n",
    "tpm_df = pd.read_csv('data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.Pancreas.gct.gz',\n",
    "                     sep='\\t', skiprows=2, usecols=pancreas_sample_ids, index_col=0)\n",
    "gene_names_s = pd.read_csv('data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_tpm.Pancreas.gct.gz',\n",
    "                            sep='\\t', skiprows=2, usecols=['Name', 'Description'], index_col=0, squeeze=True)\n",
    "print('Number of genes in table: {}'.format(tpm_df.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The table contains data for all genes in the reference annotation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-111CU-0526-SM-5EGHK</th>\n",
       "      <th>GTEX-111YS-1226-SM-5EGGJ</th>\n",
       "      <th>GTEX-1122O-0726-SM-5GIEV</th>\n",
       "      <th>GTEX-1128S-0826-SM-5GZZI</th>\n",
       "      <th>GTEX-117YX-0226-SM-5EGH6</th>\n",
       "      <th>GTEX-11DXX-0926-SM-5H112</th>\n",
       "      <th>GTEX-11EQ9-1026-SM-5H134</th>\n",
       "      <th>GTEX-11GSP-0426-SM-5A5KX</th>\n",
       "      <th>GTEX-11I78-0626-SM-5A5LZ</th>\n",
       "      <th>GTEX-11LCK-0226-SM-5A5M6</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZT9W-0926-SM-57WFS</th>\n",
       "      <th>GTEX-ZTPG-1026-SM-5DUWP</th>\n",
       "      <th>GTEX-ZV7C-0726-SM-59HKH</th>\n",
       "      <th>GTEX-ZVP2-0726-SM-59HKY</th>\n",
       "      <th>GTEX-ZVT2-2026-SM-5NQ8Q</th>\n",
       "      <th>GTEX-ZVZP-0626-SM-59HL5</th>\n",
       "      <th>GTEX-ZYFG-0826-SM-5BC5T</th>\n",
       "      <th>GTEX-ZYW4-2126-SM-59HJ9</th>\n",
       "      <th>GTEX-ZYY3-0826-SM-5E44R</th>\n",
       "      <th>GTEX-ZZPU-0726-SM-5N9C8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0546</td>\n",
       "      <td>0.04141</td>\n",
       "      <td>0.06778</td>\n",
       "      <td>0.03802</td>\n",
       "      <td>0.03170</td>\n",
       "      <td>0.01529</td>\n",
       "      <td>0.06213</td>\n",
       "      <td>0.03664</td>\n",
       "      <td>0.01931</td>\n",
       "      <td>...</td>\n",
       "      <td>0.01598</td>\n",
       "      <td>0.01906</td>\n",
       "      <td>0.06286</td>\n",
       "      <td>0.10340</td>\n",
       "      <td>0.04275</td>\n",
       "      <td>0.03054</td>\n",
       "      <td>0.07476</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.09764</td>\n",
       "      <td>0.03772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>3.0770</td>\n",
       "      <td>3.0170</td>\n",
       "      <td>5.28200</td>\n",
       "      <td>7.54500</td>\n",
       "      <td>3.42600</td>\n",
       "      <td>4.56800</td>\n",
       "      <td>4.84700</td>\n",
       "      <td>1.78200</td>\n",
       "      <td>3.65500</td>\n",
       "      <td>4.05800</td>\n",
       "      <td>...</td>\n",
       "      <td>3.91200</td>\n",
       "      <td>9.07500</td>\n",
       "      <td>6.90700</td>\n",
       "      <td>6.47800</td>\n",
       "      <td>4.18200</td>\n",
       "      <td>6.32800</td>\n",
       "      <td>5.20600</td>\n",
       "      <td>5.59100</td>\n",
       "      <td>6.64400</td>\n",
       "      <td>7.34400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>0.1564</td>\n",
       "      <td>0.1213</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.03764</td>\n",
       "      <td>0.04222</td>\n",
       "      <td>0.03521</td>\n",
       "      <td>0.03396</td>\n",
       "      <td>0.10350</td>\n",
       "      <td>0.08139</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.03549</td>\n",
       "      <td>0.12700</td>\n",
       "      <td>0.03491</td>\n",
       "      <td>0.09188</td>\n",
       "      <td>0.18990</td>\n",
       "      <td>0.06783</td>\n",
       "      <td>0.16610</td>\n",
       "      <td>0.04808</td>\n",
       "      <td>0.04338</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.01932</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02660</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.04279</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.12820</td>\n",
       "      <td>0.02703</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02667</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.05789</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.06411</td>\n",
       "      <td>0.07847</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.06105</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.05507</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.03434</td>\n",
       "      <td>0.02831</td>\n",
       "      <td>0.03726</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.03518</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 248 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   GTEX-111CU-0526-SM-5EGHK  GTEX-111YS-1226-SM-5EGGJ  \\\n",
       "Name                                                                    \n",
       "ENSG00000223972.4                    0.0000                    0.0546   \n",
       "ENSG00000227232.4                    3.0770                    3.0170   \n",
       "ENSG00000243485.2                    0.1564                    0.1213   \n",
       "ENSG00000237613.2                    0.0000                    0.0000   \n",
       "ENSG00000268020.2                    0.0000                    0.0000   \n",
       "\n",
       "                   GTEX-1122O-0726-SM-5GIEV  GTEX-1128S-0826-SM-5GZZI  \\\n",
       "Name                                                                    \n",
       "ENSG00000223972.4                   0.04141                   0.06778   \n",
       "ENSG00000227232.4                   5.28200                   7.54500   \n",
       "ENSG00000243485.2                   0.00000                   0.03764   \n",
       "ENSG00000237613.2                   0.01932                   0.00000   \n",
       "ENSG00000268020.2                   0.00000                   0.06105   \n",
       "\n",
       "                   GTEX-117YX-0226-SM-5EGH6  GTEX-11DXX-0926-SM-5H112  \\\n",
       "Name                                                                    \n",
       "ENSG00000223972.4                   0.03802                   0.03170   \n",
       "ENSG00000227232.4                   3.42600                   4.56800   \n",
       "ENSG00000243485.2                   0.04222                   0.03521   \n",
       "ENSG00000237613.2                   0.02660                   0.00000   \n",
       "ENSG00000268020.2                   0.00000                   0.00000   \n",
       "\n",
       "                   GTEX-11EQ9-1026-SM-5H134  GTEX-11GSP-0426-SM-5A5KX  \\\n",
       "Name                                                                    \n",
       "ENSG00000223972.4                   0.01529                   0.06213   \n",
       "ENSG00000227232.4                   4.84700                   1.78200   \n",
       "ENSG00000243485.2                   0.03396                   0.10350   \n",
       "ENSG00000237613.2                   0.04279                   0.00000   \n",
       "ENSG00000268020.2                   0.05507                   0.00000   \n",
       "\n",
       "                   GTEX-11I78-0626-SM-5A5LZ  GTEX-11LCK-0226-SM-5A5M6  \\\n",
       "Name                                                                    \n",
       "ENSG00000223972.4                   0.03664                   0.01931   \n",
       "ENSG00000227232.4                   3.65500                   4.05800   \n",
       "ENSG00000243485.2                   0.08139                   0.00000   \n",
       "ENSG00000237613.2                   0.12820                   0.02703   \n",
       "ENSG00000268020.2                   0.00000                   0.00000   \n",
       "\n",
       "                            ...             GTEX-ZT9W-0926-SM-57WFS  \\\n",
       "Name                        ...                                       \n",
       "ENSG00000223972.4           ...                             0.01598   \n",
       "ENSG00000227232.4           ...                             3.91200   \n",
       "ENSG00000243485.2           ...                             0.03549   \n",
       "ENSG00000237613.2           ...                             0.00000   \n",
       "ENSG00000268020.2           ...                             0.00000   \n",
       "\n",
       "                   GTEX-ZTPG-1026-SM-5DUWP  GTEX-ZV7C-0726-SM-59HKH  \\\n",
       "Name                                                                  \n",
       "ENSG00000223972.4                  0.01906                  0.06286   \n",
       "ENSG00000227232.4                  9.07500                  6.90700   \n",
       "ENSG00000243485.2                  0.12700                  0.03491   \n",
       "ENSG00000237613.2                  0.02667                  0.00000   \n",
       "ENSG00000268020.2                  0.03434                  0.02831   \n",
       "\n",
       "                   GTEX-ZVP2-0726-SM-59HKY  GTEX-ZVT2-2026-SM-5NQ8Q  \\\n",
       "Name                                                                  \n",
       "ENSG00000223972.4                  0.10340                  0.04275   \n",
       "ENSG00000227232.4                  6.47800                  4.18200   \n",
       "ENSG00000243485.2                  0.09188                  0.18990   \n",
       "ENSG00000237613.2                  0.05789                  0.00000   \n",
       "ENSG00000268020.2                  0.03726                  0.00000   \n",
       "\n",
       "                   GTEX-ZVZP-0626-SM-59HL5  GTEX-ZYFG-0826-SM-5BC5T  \\\n",
       "Name                                                                  \n",
       "ENSG00000223972.4                  0.03054                  0.07476   \n",
       "ENSG00000227232.4                  6.32800                  5.20600   \n",
       "ENSG00000243485.2                  0.06783                  0.16610   \n",
       "ENSG00000237613.2                  0.06411                  0.07847   \n",
       "ENSG00000268020.2                  0.00000                  0.00000   \n",
       "\n",
       "                   GTEX-ZYW4-2126-SM-59HJ9  GTEX-ZYY3-0826-SM-5E44R  \\\n",
       "Name                                                                  \n",
       "ENSG00000223972.4                  0.10820                  0.09764   \n",
       "ENSG00000227232.4                  5.59100                  6.64400   \n",
       "ENSG00000243485.2                  0.04808                  0.04338   \n",
       "ENSG00000237613.2                  0.00000                  0.00000   \n",
       "ENSG00000268020.2                  0.00000                  0.03518   \n",
       "\n",
       "                   GTEX-ZZPU-0726-SM-5N9C8  \n",
       "Name                                        \n",
       "ENSG00000223972.4                  0.03772  \n",
       "ENSG00000227232.4                  7.34400  \n",
       "ENSG00000243485.2                  0.00000  \n",
       "ENSG00000237613.2                  0.02640  \n",
       "ENSG00000268020.2                  0.00000  \n",
       "\n",
       "[5 rows x 248 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tpm_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The expression of a specific gene can then be queried as"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = tpm_df.loc['ENSG00000254647.2']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To facilitate queries by gene name (as long as this is unique), a gene name &rarr; ID mapping can be constructed as follows. In the GCT tables, _Name_ contains the GENCODE ID, and _Description_ the gene name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSG00000254647.2'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_to_name = gene_names_s.to_dict()\n",
    "name_to_id = {j:i for i,j in id_to_name.items()}\n",
    "name_to_id['INS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GTEX-111CU-0526-SM-5EGHK    1432.0\n",
       "GTEX-111YS-1226-SM-5EGGJ    1317.0\n",
       "GTEX-1122O-0726-SM-5GIEV    1190.0\n",
       "GTEX-1128S-0826-SM-5GZZI    3515.0\n",
       "GTEX-117YX-0226-SM-5EGH6    2810.0\n",
       "GTEX-11DXX-0926-SM-5H112     635.0\n",
       "GTEX-11EQ9-1026-SM-5H134    2744.0\n",
       "GTEX-11GSP-0426-SM-5A5KX     434.5\n",
       "GTEX-11I78-0626-SM-5A5LZ    9137.0\n",
       "GTEX-11LCK-0226-SM-5A5M6    3038.0\n",
       "Name: ENSG00000254647.2, dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g = tpm_df.loc[name_to_id['INS']]\n",
    "g.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Working with pandas DataFrames provides many useful functionalities. For example, the median expression across a set of samples can be simply calculated as"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Name\n",
       "ENSG00000223972.4    0.027255\n",
       "ENSG00000227232.4    5.553000\n",
       "ENSG00000243485.2    0.034055\n",
       "ENSG00000237613.2    0.022915\n",
       "ENSG00000268020.2    0.000000\n",
       "dtype: float64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "median_expr = tpm_df.median(axis=1)\n",
    "median_expr.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and the median expression for a specific gene can then be queried as"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2338.0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "median_expr[name_to_id['INS']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview of the cis-eQTL data\n",
    "In this section, we provide an overview of the cis-eQTL data. This consists of the following set of files (available from the GTEx Portal), for each tissue:\n",
    "* Gene/eGene summary statistics<br>(files ending in *.egenes.txt.gz)\n",
    "* The list of all significant variant-gene associations found in a tissue<br>(files ending in *.signif_variant_gene_pairs.txt.gz)\n",
    "* The association results for all variant-gene pairs tested in a tissue<br>(files ending in *.allpairs.txt.gz)\n",
    "\n",
    "Additionally, all input files used for eQTL mapping are available from the Portal, including normalized expression and covariates (both described in the tutorial).\n",
    "\n",
    "Here, we will focus on the eGene summary and significant pair tables.\n",
    "\n",
    "Begin by loading the gene summary table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df = pd.read_csv('data/Pancreas.v7.egenes.txt.gz', sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To view the different columns available in this table, inspect a specific row:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name                                  INS\n",
       "gene_chr                                    11\n",
       "gene_start                             2181009\n",
       "gene_end                               2182571\n",
       "strand                                       -\n",
       "num_var                                   9167\n",
       "beta_shape1                            1.04048\n",
       "beta_shape2                            1457.85\n",
       "true_df                                163.574\n",
       "pval_true_df                       0.000846469\n",
       "variant_id                  11_2310683_G_A_b37\n",
       "tss_distance                            128112\n",
       "chr                                         11\n",
       "pos                                    2310683\n",
       "ref                                          G\n",
       "alt                                          A\n",
       "num_alt_per_site                             1\n",
       "rs_id_dbSNP147_GRCh37p13           rs117595442\n",
       "minor_allele_samples                        11\n",
       "minor_allele_count                          11\n",
       "maf                                      0.025\n",
       "ref_factor                                   1\n",
       "pval_nominal                       0.000415083\n",
       "slope                                -0.567289\n",
       "slope_se                              0.157741\n",
       "pval_perm                             0.697075\n",
       "pval_beta                             0.693643\n",
       "qval                                  0.331348\n",
       "pval_nominal_threshold             1.62118e-05\n",
       "Name: ENSG00000254647.2, dtype: object"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.loc[name_to_id['INS']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As this example shows, the table contains data for all genes, not just eGenes. To select eGenes at FDR ≤ 0.05, corresponding to the set used in GTEx Consortium analyses, run:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of eGenes: 7146\n"
     ]
    }
   ],
   "source": [
    "egenes_df = gene_df[gene_df['qval']<=0.05]\n",
    "print('Number of eGenes: {}'.format(egenes_df.shape[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As a toy example, we'll next show how to extract all significant variant-gene pairs corresponding to the eGene with the strongest association in Pancreas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>pval_nominal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000197728.5</th>\n",
       "      <td>12_56435929_C_G_b37</td>\n",
       "      <td>3.243440e-97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000013573.12</th>\n",
       "      <td>12_31225344_A_G_b37</td>\n",
       "      <td>1.065180e-93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000166435.11</th>\n",
       "      <td>11_74659302_A_T_b37</td>\n",
       "      <td>4.150050e-84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000174652.13</th>\n",
       "      <td>19_9517537_C_G_b37</td>\n",
       "      <td>1.265260e-82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000204792.2</th>\n",
       "      <td>2_75185856_G_T_b37</td>\n",
       "      <td>5.738330e-80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             variant_id  pval_nominal\n",
       "gene_id                                              \n",
       "ENSG00000197728.5   12_56435929_C_G_b37  3.243440e-97\n",
       "ENSG00000013573.12  12_31225344_A_G_b37  1.065180e-93\n",
       "ENSG00000166435.11  11_74659302_A_T_b37  4.150050e-84\n",
       "ENSG00000174652.13   19_9517537_C_G_b37  1.265260e-82\n",
       "ENSG00000204792.2    2_75185856_G_T_b37  5.738330e-80"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "egenes_df[['variant_id', 'pval_nominal']].sort_values('pval_nominal').head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the association results for all significant variant-gene pairs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df = pd.read_csv('data/Pancreas.v7.signif_variant_gene_pairs.txt.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, select the pairs corresponding to the top gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_pairs_df = pairs_df[pairs_df['gene_id']=='ENSG00000197728.5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "      <th>min_pval_nominal</th>\n",
       "      <th>pval_beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>486740</th>\n",
       "      <td>12_56271219_T_A_b37</td>\n",
       "      <td>ENSG00000197728.5</td>\n",
       "      <td>-164418</td>\n",
       "      <td>42</td>\n",
       "      <td>42</td>\n",
       "      <td>0.095454</td>\n",
       "      <td>2.491330e-05</td>\n",
       "      <td>-0.687914</td>\n",
       "      <td>0.159012</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>3.243440e-97</td>\n",
       "      <td>9.644380e-89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486741</th>\n",
       "      <td>12_56354272_G_C_b37</td>\n",
       "      <td>ENSG00000197728.5</td>\n",
       "      <td>-81365</td>\n",
       "      <td>115</td>\n",
       "      <td>140</td>\n",
       "      <td>0.331754</td>\n",
       "      <td>8.268610e-14</td>\n",
       "      <td>0.681209</td>\n",
       "      <td>0.084259</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>3.243440e-97</td>\n",
       "      <td>9.644380e-89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486742</th>\n",
       "      <td>12_56364321_A_G_b37</td>\n",
       "      <td>ENSG00000197728.5</td>\n",
       "      <td>-71316</td>\n",
       "      <td>122</td>\n",
       "      <td>149</td>\n",
       "      <td>0.341743</td>\n",
       "      <td>1.424730e-17</td>\n",
       "      <td>0.721492</td>\n",
       "      <td>0.076173</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>3.243440e-97</td>\n",
       "      <td>9.644380e-89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486743</th>\n",
       "      <td>12_56366160_G_C_b37</td>\n",
       "      <td>ENSG00000197728.5</td>\n",
       "      <td>-69477</td>\n",
       "      <td>48</td>\n",
       "      <td>54</td>\n",
       "      <td>0.123288</td>\n",
       "      <td>2.009430e-05</td>\n",
       "      <td>0.580996</td>\n",
       "      <td>0.132702</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>3.243440e-97</td>\n",
       "      <td>9.644380e-89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486744</th>\n",
       "      <td>12_56367966_C_T_b37</td>\n",
       "      <td>ENSG00000197728.5</td>\n",
       "      <td>-67671</td>\n",
       "      <td>40</td>\n",
       "      <td>45</td>\n",
       "      <td>0.102740</td>\n",
       "      <td>1.170990e-05</td>\n",
       "      <td>-0.577004</td>\n",
       "      <td>0.128026</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>3.243440e-97</td>\n",
       "      <td>9.644380e-89</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 variant_id            gene_id  tss_distance  ma_samples  \\\n",
       "486740  12_56271219_T_A_b37  ENSG00000197728.5       -164418          42   \n",
       "486741  12_56354272_G_C_b37  ENSG00000197728.5        -81365         115   \n",
       "486742  12_56364321_A_G_b37  ENSG00000197728.5        -71316         122   \n",
       "486743  12_56366160_G_C_b37  ENSG00000197728.5        -69477          48   \n",
       "486744  12_56367966_C_T_b37  ENSG00000197728.5        -67671          40   \n",
       "\n",
       "        ma_count       maf  pval_nominal     slope  slope_se  \\\n",
       "486740        42  0.095454  2.491330e-05 -0.687914  0.159012   \n",
       "486741       140  0.331754  8.268610e-14  0.681209  0.084259   \n",
       "486742       149  0.341743  1.424730e-17  0.721492  0.076173   \n",
       "486743        54  0.123288  2.009430e-05  0.580996  0.132702   \n",
       "486744        45  0.102740  1.170990e-05 -0.577004  0.128026   \n",
       "\n",
       "        pval_nominal_threshold  min_pval_nominal     pval_beta  \n",
       "486740                0.000041      3.243440e-97  9.644380e-89  \n",
       "486741                0.000041      3.243440e-97  9.644380e-89  \n",
       "486742                0.000041      3.243440e-97  9.644380e-89  \n",
       "486743                0.000041      3.243440e-97  9.644380e-89  \n",
       "486744                0.000041      3.243440e-97  9.644380e-89  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Significant variant-gene pairs for gene RPS26: 155\n"
     ]
    }
   ],
   "source": [
    "print('Significant variant-gene pairs for gene {}: {}'.format(\n",
    "    id_to_name['ENSG00000197728.5'], selected_pairs_df.shape[0]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
